### ISSUE TITLE
[exporter/loadbalancing] Data loss still unavoidable under high load and autoscaling

### ISSUE THREAD

**Issue by jamesmoessis** [open-telemetry/opentelemetry-collector-contrib#36717]
### Component(s)

exporter/loadbalancing

## Description

We had the previous issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35378 

However this is unfortunately not quite solved.

There is the top-level exporter queue and the sub exporter queues. To avoid data loss on an autoscale down event, there can't be queues on the sub-exporter. However if you disable this queue, it's extremely slow because the endpoints are exported to synchronously in-series. 

So, basically, if we have a queue, we get data loss on scaling. If we don't have a queue, we get data loss from performance issues because it's too slow to export. This problem gets worse the more downstream you have.

I've annotated the problematic code here with comments: 

```go
for exp, td := range exporterSegregatedTraces {
  start := time.Now()
  // with no queue, this blocks.
  // with queue, this doesn't block but causes data loss when the downstream gets scaled down.
  err := exp.ConsumeTraces(ctx, td) 
  exp.consumeWG.Done()
  errs = multierr.Append(errs, err)
  ...
}
```

## Steps to Reproduce

Disable sub-exporter queue.
```yaml
  loadbalancing:
    # top-level exporter queue
    sending_queue:
      queue_size: 6000
      num_consumers: 200 # even with high consumer amount the problem persisted for us
    protocol:
      otlp:
        sending_queue:
          enabled: false
```

Then throw some load at the collector. 

Result is severe performance issues and high export times.

## Proposed solution

I think we need to have a way of disabling the sub exporter queue without blocking in series here. An `errgroup` or similar so the consumers can export to different endpoints in-parallel without relying on the sub exporter queue. Open to other ideas to solve the problem, but this seems simplest to me.

### Collector version

v0.115.0

**Comment by jamesmoessis** [comment]
cc @MovieStoreGuy @Fiery-Fenix

**Comment by jpkrohling** [comment]
I wonder if it's time to rethink the way we are handling failures at the load-balancing exporter and delegate some of that to the failoverconnector.

**Comment by jamesmoessis** [comment]
@jpkrohling the failoverconnector idea is interesting, I'd be interested to hear more about that. 

I think to solve this would be relatively easy though - if there is no queue on the sub exporter we should just fork and wait the exporters with using `errgroup` or similar. 

I also would like to solve the problem of when a single sub-exporter fails, the whole batch is retried instead of just the failed parts. Would this be solved by something like failover connector?

**Comment by jamesmoessis** [comment]
@jpkrohling I did actually start to work on this, but stumbled upon an issue which I found to be troubling.

There's no way to classify an error, or if one of the exporters fails, then all of them will retry. In other words, say you have 100 sub-exporters, and one of them fails - then all 100 exporters will be re-sent the data to export. We need some way around this problem too. 

https://github.com/open-telemetry/opentelemetry-collector/issues/7047 would potentially solve this as it could potentially introduce a more fine grained error handling into exporters.

Or, there could be other ways to approach this specifically in the load balancing exporter

**Comment by jamesmoessis** [comment]
Our current solution to this is to not autoscale our stateful samplers. And then, disable all queueing and retrying on the top level exporter. 

With the current error handling in the collector, it's not really possible for the top level exporter to distinguish what has errored and what hasn't, and so the top level retrier over-compensates when it receives an error, retrying all data rather than just failed

**Comment by snuggie12** [comment]
If it helps I wanted to provide my experience. I run the k8s resolver with `return_hostnames: true` (the feature I wrote to get istio working.) If I only restart the samplers I have no issues. However, if I restart both the collectors (with the lbexporter) and the samplers at the same time I have very inconsistent results.

In my test environment I have 3 samplers and 2 collectors. Due to the restart policy, `otel-sampler-2` never has an issue since it restarts before the collectors. However, everything else is completely intermittent. Most of the time both `otel-sampler-0` and `otel-sampler-1` cannot be reached. After about 2-5 minutes either 1 or both of the collectors will recover. However, some times one collector recovers both samplers, sometimes only 1 collector recovers both samplers. However, there have been plenty of tests where one collector has zero issues and the other one has issues.

With exporter resiliency settings disabled they never recover. I have disabled subexporter resiliency settings and it doesn't seem to change anything aside from noisy log lines telling me to turn the settings on.

Though this is anecdotal, while I was developing `return_hostnames` I don't recall ever having this issue. Given that the loadbalancing exporter has changed, I have to ponder if the problem is in the exporterhelper or grpc. Based on what I read, it seemed like `pick_first` would be better than `round_robin`, but given that there's only 1 IP to look at it feels like this shouldn't matter and tests have proven that out.

However, what I can't figure out is why there's always a consistency where the the channel and subchannels always exit idle mode, discovers the address, states it's `READY` and the very next log is errors saying spans have to be dropped. And then one the ones that fix it's error'ing, enters `IDLE` and it's magically fixed.

In summary, I've run the same tests about 20 times and the results are always a random combination across the following elements:
- Whether 0, 1, or 2 collectors recover
- If one does recover, how long it takes to recover
- Whether `otel-sampler-0`, `otel-sampler-1`, neither, or both are unreachable

In general my thought has been to minimize resiliency settings. If the subexporters have super small queues then things should move to the load balancer which should have the new IPs and create new subexporters as fast as possible. However, it seems I don't understand how the code works because it doesn't seem to work that way.

**Comment by jamesmoessis** [comment]
@snuggie12 thanks for sharing your experience. Have you tried disabling the top-level exporter queue and only enabling the sub-level one? I believe the top-level queue is relatively new. That's how we configure it in ours (granted, we use the static resolver since the dynamic resolvers we find to be unreliable, especially in the scenario of a downstream being rescheduled).

**Comment by alexchowle** [comment]
@jamesmoessis I just wanted to thank you for sticking with this issue. I have exactly the same issue when losing downstream Tail Sampling Collectors. We accept some Span loss, but it's frustrating that Pod rescheduling/cluster maintenance/deployments/scale downs cause such loss.

**Comment by jamesmoessis** [comment]
No worries @alexchowle, it's a tricky problem. You can avoid data loss by using the static resolver and disabling top-level queue/retry. But that is of course at the cost of scaling your downstream.

**Comment by snuggie12** [comment]
> [@snuggie12](https://github.com/snuggie12) thanks for sharing your experience. Have you tried disabling the top-level exporter queue and only enabling the sub-level one? I believe the top-level queue is relatively new. That's how we configure it in ours (granted, we use the static resolver since the dynamic resolvers we find to be unreliable, especially in the scenario of a downstream being rescheduled).

@jamesmoessis Nothing has been settled yet, but currently things are working with both queues on, though in my tests the main exporter queue is not used.

I noticed from a comment that you or someone else made that `static` was working but was making it impossible to autoscale. ~~At some point something must have changed either in the exporter helper or, more likely, the grpc library. The reason I made the feature in the k8s resolver to return hostnames was because `static` would not re-resolve DNS (in fact I think the docs still say only one resolution happens.)~~

UPDATE: I moved my candidate config to another environment with more scale and the behavior of everything is radically different. At this point I feel like I need to make a separate issue since it no longer feels relevant.

So now there's some sort of ~5m timeout that happens despite me setting the retry settings for lower. In my lower environments the queue doesn't fill up that fast so I'm going to test in envs with more traffic to see how it works.

Has anyone tried the persistent storage queue? Would that survive sampler's being restarted?

**Comment by snuggie12** [comment]
Am I correct that `otelcol_exporter_send_failed_spans_total` represents one `attempt` failing to send? That is, if something tried to send a span, failed, failed a retry, but then succeeded on the 3rd try (2nd retry) that the value of this metric would be `2`?

Based on a recent graph I was looking at on the receiver end I see an average rate of 400 accepted and then upon one of the samplers restarting it dips to about 300 with only 1 out of my 3 collectors having an issue sending to that sampler and then 5 minutes later the queue gets flushed out and jumps up to a brief moment of 1400.

If that's the case I think I'm alright since the queue never gets full on the subexporter and with `return_hostnames` turned on the endpoints never change.

@jamesmoessis it won't help with scale down, but have you tried `return_hostnames` on with the k8s resolver? Should help with other issues like the IP address changing.
